src/v/kafka/server/connection_context.cc

```diff
@@ -64,24 +64,77 @@ class invalid_virtual_connection_id : public std::exception {
     explicit invalid_virtual_connection_id(ss::sstring msg)
       : _msg(std::move(msg)) {}
 
+    const char* what() const noexcept final { return _msg.c_str(); }
+
 private:
     ss::sstring _msg;
 };
 
-bytes parse_virtual_connection_id(const kafka::request_header& header) {
+// Tuple containing virtual connection id and client id. It is returned after
+// parsing parts of virtual connection id.
+struct virtual_connection_client_id {
+    virtual_connection_id v_connection_id;
+    std::optional<std::string_view> client_id;
+};
+
+vcluster_connection_id parse_vcluster_connection_id(std::string_view str) {
+    vcluster_connection_id cid;
+    std::stringstream sstream(str.data());
+    sstream >> std::hex >> cid;
+    return cid;
+}
+
+/**
+ * Virtual connection id is encoded as with the following structure:
+ *
+ * [vcluster_id][connection_id][actual client id]
+ *
+ * vcluster_id - is a string encoded XID representing virtual cluster (20
+ *               characters)
+ * connection_id - is a hex encoded 32 bit integer representing virtual
+ *                 connection id (8 characters)
+ *
+ * client_id - standard protocol defined client id
+ */
+virtual_connection_client_id
+parse_virtual_connection_id(const kafka::request_header& header) {
+    static constexpr size_t connection_id_str_size
+      = sizeof(vcluster_connection_id::type) * 2;
+    static constexpr size_t v_connection_id_size = xid::str_size
+                                                   + connection_id_str_size;
     if (header.client_id_buffer.empty()) {
         throw invalid_virtual_connection_id(
           "virtual connection client id can not be empty");
     }
 
-    // TODO: should we use vcluster_id here ?
-    return bytes{
-      reinterpret_cast<const uint8_t*>(header.client_id_buffer.begin()),
-      header.client_id_buffer.size()};
+    if (header.client_id->size() < v_connection_id_size) {
+        throw invalid_virtual_connection_id(fmt::format(
+          "virtual connection client id size must contain at least {} "
+          "characters. Current size: {}",
+          v_connection_id_size,
+          header.client_id_buffer.size()));
+    }
+
+    virtual_connection_id connection_id{
+      .virtual_cluster_id = xid::from_string(
+        std::string_view(header.client_id->begin(), xid::str_size)),
```

Comment:
Will an error be raised if the header doesn't contain a valid `xid` string?

Reply:
yes, added a test validating this